This patch is based on the upstream commit described below, adapted for use in
the Debian package by Peter Michael Green.

commit bb133c171dbaebf20858e7073a78491994d59b45
Author: Ian Douglas Scott <idscott@system76.com>
Date:   Fri Sep 22 18:11:43 2023 -0700

    server: Use `std::fs` APIs instead of `nix` where possible

--- wayland-server/src/socket.rs
+++ wayland-server/src/socket.rs
@@ -10,3 +10,6 @@
     os::unix::{
-        io::{AsRawFd, FromRawFd, RawFd},
+        fs::OpenOptionsExt,
+    },
+    os::unix::{
+        io::{AsRawFd, RawFd},
         net::{UnixListener, UnixStream},
@@ -20,5 +20,1 @@
-use nix::{
-    fcntl::{flock, open, FlockArg, OFlag},
-    sys::stat::{lstat, Mode},
-    unistd::unlink,
-};
+use nix::fcntl::{flock, FlockArg};
@@ -83,9 +82,7 @@
-            let lock_fd = open(
-                &lock_path,
-                OFlag::O_CREAT | OFlag::O_CLOEXEC | OFlag::O_RDWR,
-                Mode::S_IRUSR | Mode::S_IWUSR | Mode::S_IRGRP | Mode::S_IWGRP,
-            )
-            .map_err(|_| BindError::PermissionDenied)?;
-
-            // SAFETY: We have just opened the file descriptor.
-            _lock = unsafe { File::from_raw_fd(lock_fd) };
+            _lock = File::options()
+                .create(true)
+                .read(true)
+                .write(true)
+                .mode(0o660)
+                .open(&lock_path)
+                .map_err(|_| BindError::PermissionDenied)?;
@@ -100,1 +100,1 @@
-            if flock(lock_fd, FlockArg::LockExclusiveNonblock).is_err() {
+            if flock(_lock.as_raw_fd(), FlockArg::LockExclusiveNonblock).is_err() {
@@ -120,17 +117,17 @@ impl ListeningSocket {
         }
 
         // check if an old socket exists, and cleanup if relevant
-        match lstat(&socket_path) {
-            Err(nix::Error::ENOENT) => {
+        match socket_path.try_exists() {
+            Ok(false) => {
                 // none exist, good
             }
-            Ok(_) => {
+            Ok(true) => {
                 // one exist, remove it
-                unlink(&socket_path).map_err(|_| BindError::AlreadyInUse)?;
+                fs::remove_file(&socket_path).map_err(|_| BindError::AlreadyInUse)?;
             }
             Err(e) => {
                 // some error stat-ing the socket?
-                return Err(BindError::Io(e.into()));
+                return Err(BindError::Io(e));
             }
         }
 
@@ -191,8 +188,8 @@ impl AsFd for ListeningSocket {
 
 impl Drop for ListeningSocket {
     fn drop(&mut self) {
-        let _ = unlink(&self.socket_path);
-        let _ = unlink(&self.lock_path);
+        let _ = fs::remove_file(&self.socket_path);
+        let _ = fs::remove_file(&self.lock_path);
     }
 }
 
